agent/consul/federation_state_endpoint.go

```diff
@@ -0,0 +1,182 @@
+package consul
+
+import (
+	"fmt"
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/structs"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+// FederationState endpoint is used to manipulate federation states from all
+// datacenters.
+type FederationState struct {
+	srv *Server
+	// TODO(rb): does this need a Logger?
+}
+
+func (c *FederationState) Apply(args *structs.FederationStateRequest, reply *bool) error {
+	// Ensure that all federation state writes go to the primary datacenter. These will then
+	// be replicated to all the other datacenters.
+	args.Datacenter = c.srv.config.PrimaryDatacenter
+
+	if done, err := c.srv.forward("FederationState.Apply", args, args, reply); done {
+		return err
+	}
+	defer metrics.MeasureSince([]string{"federation_state", "apply"}, time.Now())
+
+	// Fetch the ACL token, if any.
+	rule, err := c.srv.ResolveToken(args.Token)
+	if err != nil {
+		return err
+	}
+	if rule != nil && rule.OperatorWrite(nil) != acl.Allow {
+		return acl.ErrPermissionDenied
+	}
+
+	if args.State == nil || args.State.Datacenter == "" {
+		return fmt.Errorf("invalid request: missing federation state datacenter")
+	}
+
+	switch args.Op {
+	case structs.FederationStateUpsert:
+		if !c.srv.router.HasDatacenter(args.State.Datacenter) &&
+			!c.srv.config.DisableFederationStateDatacenterNameValidation {
+			return fmt.Errorf("cannot upsert federation state for datacenter that does not exist: %s", args.State.Datacenter)
```

Comment:
Is it worth commenting why this is important to enforce? Given that you need to have operator:write already, you could do a lot worse than fill up the state store with arbitrary Fed States for non-existent DCs. I can imagine there could be ordering issues where this would at least cause errors for a while until things resolve (but might be wrong about that).

Just wondering if there are other strong reasons to enforce this I'm missing?

Reply:
Well if there was a strong reason I probably would've written a test. I commented this validation out and everything important passed so I can't remember why I thought this was important to enforce.